scripts/backout_data_collection.py

```diff
@@ -0,0 +1,169 @@
+    filtered_list = []
+    counter = 0
+
+    pbar = tqdm(total=count_limit, desc="Filtering commits")
+
+    for commit in repository.get_commits(
+        include_no_bug=False, include_backouts=True, include_ignored=False
+    ):
+        # add commit if it was backed out and the bug is fixed
+        if len(commit["backedoutby"]) > 0 and is_bug_fixed(
+            bug_id=commit["bug_id"], bug_dict=bug_dict
+        ):
+            fixing_commit = find_next_commit(
+                commit["bug_id"], bug_to_commit_dict, commit["node"]
+            )
+
+            # if fixing commit could not be found or is another backing out commit, do not add it to dataset
+            if (
+                fixing_commit["node"] == commit["backedoutby"]
+                or len(fixing_commit["backsout"]) > 0
+            ):
+                counter += 1
+                continue
+
+            # add the hashes of the bug-inducing commit, the back out commit, and the fixing commit
+            # include metadata such as push date and description for further context
+            list_entry = {
+                "bug_id": commit["bug_id"],
+                "inducing_commit": {
+                    "node": commit["node"],
+                    "pushdate": commit["pushdate"],
+                    "desc": commit["desc"],
+                },
+                "backout_commit": {
+                    "node": commit["backedoutby"],
+                    "pushdate": commit_dict[commit["backedoutby"]]["pushdate"],
+                    "desc": commit_dict[commit["backedoutby"]]["desc"],
+                },
+                "fixing_commit": {
+                    "node": fixing_commit["node"],
+                    "pushdate": fixing_commit["pushdate"],
+                    "desc": fixing_commit["desc"],
+                },
+            }
+
+            filtered_list.append(list_entry)
+            counter += 1
+            pbar.update(1)
+
+            if counter >= count_limit:
+                break
+
+    json_data = json.dumps(filtered_list, indent=4)
+
+    if not os.path.exists(directory_path):
+        os.makedirs(directory_path)
+        print(f"Directory {directory_path} created")
+
+    with open(destination_filepath, "w") as file:
+        file.write(json_data)
+
+    logger.info(f"Data successfully saved to {destination_filepath}")
+
+    return
+
+
+def find_next_commit(bug_id: int, bug_to_commit_dict: dict, inducing_node: str) -> dict:
+    inducing_commit_found = False
+    for commit in bug_to_commit_dict[bug_id]:
+        # if the inducing commit has been found, find next commit that has not been backed out
+        if inducing_commit_found:
+            if len(commit["backedoutby"]) == 0:
+                return commit
+
+        if commit["node"] == inducing_node:
+            inducing_commit_found = True
+
+    return commit
+
+
+def main():
+    download_databases()
+
+    commit_dict, bug_to_commit_dict, bug_dict = preprocess_commits_and_bugs()
```

Comment:
We may want to consider the space complexity when iterating over the whole dataset.

Reply:
Removed unused keys when constructing the dictionaries and implemented a cache to use generated dictionaries from previous instances of running the code via saving them as JSON files. Let me know if this needs additional changes/fixes!